libutils/threaded_queue.c

```diff
@@ -0,0 +1,393 @@
+/*
+   Copyright 2018 Northern.tech AS
+
+   This file is part of CFEngine 3 - written and maintained by Northern.tech AS.
+
+   This program is free software; you can redistribute it and/or modify it
+   under the terms of the GNU General Public License as published by the
+   Free Software Foundation; version 3.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#include <alloc.h>
+#include <logging.h>
+#include <mutex.h>
+#include <pthread.h>
+#include <threaded_queue.h>
+
+
+#define EXPAND_FACTOR     2
+#define DEFAULT_CAPACITY 16
+
+/** @struct ThreadedQueue_
+  @brief An implementation of a thread safe queue based on a circular array
+
+  Can enqueue, dequeue and give various statistics about its contents, like
+  amount of elements, capacity and if it is empty. Has the ability to block
+  if queue is empty, waiting for new elements to be queued.
+  */
+struct ThreadedQueue_ {
+    pthread_mutex_t *lock;            /**< Thread lock for accessing data. */
+    pthread_cond_t *pcond;            /**< Thread condition for blocking.  */
+    void (*ItemDestroy) (void *item); /**< Data-specific destroy function. */
+    void **data;                      /**< Internal array of elements.     */
+    size_t head;                      /**< Current position in queue.      */
+    size_t tail;                      /**< Current end of queue.           */
+    size_t size;                      /**< Current size of queue.          */
+    size_t capacity;                  /**< Current memory allocated.       */
+};
+
+static void _DestroyRange(ThreadedQueue *queue, size_t start, size_t end);
+static void _ExpandIfNecessary(ThreadedQueue *queue);
+
+static inline void *_PopNonLock(ThreadedQueue *queue);
+static inline void _PushNonLock(ThreadedQueue *queue, void *item);
```

Comment:
I think we don't need these two functions (and a "mode" of the structure that doesn't care about the locks). And they should not start with an underscore. They are `static`, that says it all.